proxy/src/transport/tls/config.rs

```diff
@@ -66,6 +75,33 @@ pub struct ServerConfig(pub(super) Arc<rustls::ServerConfig>);
 pub type ClientConfigWatch = Watch<Option<ClientConfig>>;
 pub type ServerConfigWatch = Watch<Option<ServerConfig>>;
 
+/// The configuration in effect for a client (`ClientConfig`) or server
+/// (`ServerConfig`) TLS connection.
+#[derive(Clone, Debug)]
+pub struct ConnectionConfig<C> where C: Clone + std::fmt::Debug {
+    pub identity: Identity,
+    pub config: C,
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum ReasonForNoTls {
+    /// TLS is disabled.
+    Disabled,
+
+    /// TLS was enabled but the configuration isn't available (yet).
+    NoConfig,
+
+    /// TLS isn't implemented for the connection between the proxy and the
+    /// control plane yet.
+    NotImplementedForControlPlane,
+
+    /// TLS is only enabled for HTTP (HTTPS) right now.
+    NotImplementedForNonHttp,
```

Comment:
Do we expect these individual `NotImplemented...` variants to be exposed to the user? Otherwise, especially if we intend to implement some of these cases soon, we might want to just have one `NotImplemented` case.

Reply:
I would like to expose the reasoning for not using TLS in more detail than what I've done here. In particular, I want to replace `NoConfig` with more specific reasoning like `CaBundleIsMissing`, `PrivateKeyDoesNotMatchCertificate`, etc., so that it serves as a more useful diagnostic tool for likely problems that the proxy will encounter. I only folded them into one case for telemetry reporting purposes in this PR due to lack of time and also because I don't know what would need to be changed in the telemetry system to support that kind of change.

Note that the "no TLS for the control plane connection" case will go away very soon.

Reply:
> I would like to expose the reasoning for not using TLS in more detai

Agreed, I'm in support of this potentially having a lot of variants if necessary.